From: "Barak A. Pearlmutter" <barak+git@pearlmutter.net>
Date: Thu, 17 Sep 2015 15:29:35 +0100
Subject: documentation typos

---
 CHANGES | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/CHANGES b/CHANGES
index 824e105..0fce952 100644
--- a/CHANGES
+++ b/CHANGES
@@ -30,7 +30,7 @@ Version 1.41
 
     Contributors:
         Kevin Cozens, Gordon McNutt, Doug Currie, Sean Long, Tim Cas, Joey,
-        and Richard Copley, and CMarinier.
+        Richard Copley, and CMarinier.
 
 Version 1.40
     Bugs fixed:
