From: "Barak A. Pearlmutter" <barak+git@cs.nuim.ie>
Date: Mon, 18 Feb 2013 09:01:03 +0000
Subject: scheme.h

---
 scheme.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/scheme.h b/scheme.h
index 6f145f1..1ef4c43 100644
--- a/scheme.h
+++ b/scheme.h
@@ -70,7 +70,7 @@
 #endif
 
 #ifndef USE_PLIST
-# define USE_PLIST 0
+# define USE_PLIST 1
 #endif
 
 /* To force system errors through user-defined error handling (see *error-hook*) */
@@ -95,7 +95,7 @@
 #endif
 
 #ifndef INLINE
-# define INLINE
+# define INLINE inline
 #endif
 
 #ifndef USE_INTERFACE
